File: app/src/main/java/pl/piotrskiba/dailywallpaper/MainViewModel.java
package pl.piotrskiba.dailywallpaper;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.annotation.NonNull;

import java.util.List;

import pl.piotrskiba.dailywallpaper.database.AppDatabase;
import pl.piotrskiba.dailywallpaper.database.ImageEntry;

public class MainViewModel extends AndroidViewModel {

    private LiveData<List<ImageEntry>> favoriteImages;

    public MainViewModel(@NonNull Application application) {
        super(application);

        AppDatabase database = AppDatabase.getInstance(getApplication());
        favoriteImages = database.imageDao().loadAllImages();
    }

    public LiveData<List<ImageEntry>> getFavoriteImages(){
        return favoriteImages;
    }
}

File: app/src/main/java/pl/piotrskiba/dailywallpaper/MainViewModel.kt
package pl.piotrskiba.dailywallpaper

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import pl.piotrskiba.dailywallpaper.interfaces.ImageListLoadedListener
import pl.piotrskiba.dailywallpaper.models.ImageList
import pl.piotrskiba.dailywallpaper.utils.NetworkUtils

class MainViewModel(application: Application) : AndroidViewModel(application) {
    private val context = getApplication<Application>().applicationContext

    private var allImages: MutableLiveData<ImageList>? = null
    private var categoryImages: ArrayList<MutableLiveData<ImageList>?> = ArrayList()

    init{
        val allCategories = context.resources.getStringArray(R.array.pref_category_values)

        repeat(allCategories.size) {
            categoryImages.add(null)
        }
    }

    fun getAllImages(): LiveData<ImageList> {
        allImages ?: run {
            allImages = MutableLiveData()
            loadAllImages()
        }

        return allImages!!
    }

    fun getCategoryImages(category: Int): LiveData<ImageList> {

        categoryImages[category] ?: run {
            categoryImages[category] = MutableLiveData()
            loadCategoryImages(category)
        }

        return categoryImages[category]!!
    }

    private fun loadAllImages() {
        NetworkUtils.getAllImages(context, object : ImageListLoadedListener{
            override fun onImageListLoaded(result: ImageList) {
                allImages?.value = result
            }

            override fun onImageListLoadingError() {
                allImages?.value = null
            }

        })
    }

    private fun loadCategoryImages(category: Int) {
        NetworkUtils.getCategoryImages(context, category, object : ImageListLoadedListener{
            override fun onImageListLoaded(result: ImageList) {
                categoryImages[category]?.value = result
            }

            override fun onImageListLoadingError() {
                categoryImages[category]?.value = null
            }

        })
    }
}